### PR TITLE
Brotli import error #4697

### PR DESCRIPTION
**Pull request by Anupam-USP**
I tried to solve #4697 but if there is some problem please do let me know. Also this is my first pull request so if there is any formatting error, please correct me!

### CODE REVIEW

File: scrapy/downloadermiddlewares/httpcompression.py [hunk 1]
```diff
@@ -73,7 +75,11 @@ def _decode(self, body, encoding):
                 # http://www.gzip.org/zlib/zlib_faq.html#faq38
                 body = zlib.decompress(body, -15)
         if encoding == b'br' and b'br' in ACCEPTED_ENCODINGS:
-            body = brotli.decompress(body)
+            try:
+                body = brotli.decompress(body)
+            except ImportError:
```
Comment (wRAR) [inline scrapy/downloadermiddlewares/httpcompression.py:80]: This exception cannot happen here as far as I know, can it?

File: scrapy/downloadermiddlewares/httpcompression.py [hunk 2]
```diff
@@ -8,12 +9,13 @@
...
 
 ACCEPTED_ENCODINGS = [b'gzip', b'deflate']
+logger = logging.getLogger(__name__)
 
 try:
     import brotli
     ACCEPTED_ENCODINGS.append(b'br')
 except ImportError:
-    pass
+    logger.error("brotli not installed")
```
Comment (wRAR) [inline scrapy/downloadermiddlewares/httpcompression.py:18]: We shouldn't log an error here as this is not a problem

**Review by wRAR — CHANGES_REQUESTED** [review]

File: scrapy/downloadermiddlewares/httpcompression.py (hunk 2 above)
Comment (Anupam-USP) [reply scrapy/downloadermiddlewares/httpcompression.py:18]: Ok I thought this is an Import error, should I log a warning then?

Comment (wRAR) [reply scrapy/downloadermiddlewares/httpcompression.py:18]: We still support configurations without brotli, please read the description of the needed changes in the original issue.

File: scrapy/downloadermiddlewares/httpcompression.py (hunk 1 above)
Comment (Anupam-USP) [reply scrapy/downloadermiddlewares/httpcompression.py:80]: I think here I have to make changes here and leave rest as it is, and if not installed than give a warning? Will it do the work?

Comment (wRAR) [reply scrapy/downloadermiddlewares/httpcompression.py:80]: Again, please read the instructions in the issue:

> If the user sets br manually on the Accept-Encoding header, I think we should:
> 
> * Log a warning if brotlipy is not installed.
> * Have the middleware cause an error if the server sends the response br-encoded, and brotlipy is not installed.

So you need to raise an error here, it brotli is not installed, and in addition to this you need to log a warning if brotli is not installed but ACCEPTED_ENCODINGS contains 'br' (not sure which is the best place for this).

Comment (Anupam-USP) [reply scrapy/downloadermiddlewares/httpcompression.py:80]: Screenshot 2021-03-12 201909

I think this much change can handle the issue?

Comment (wRAR) [reply scrapy/downloadermiddlewares/httpcompression.py:80]: As we are already catching ImportError in this file, you can just set a flag there and check it here.

Also, I'm not sure whether logging an error and returning the compressed body or raising an exception instead is a better way to handle this.

**Comment by Anupam-USP** [comment]
I am not able to push my change further, it is showing that `This branch is 3 commits ahead, 6 commits behind scrapy:master.` Should I close this pull request and create a new one or is there any other way? I have tried [this](https://stackoverflow.com/questions/41283955/github-keeps-saying-this-branch-is-x-commits-ahead-y-commits-behind) solution.

**Comment by wRAR** [comment]
No, you shouldn't need to make new PRs if you just created additional commits in the same branch.

**Comment by Anupam-USP** [comment]
Actually I tried pushing with my master and it is recognized in #4697 conversation. In this branch I got the above error and not able to push

**Comment by Anupam-USP** [comment]
I am really sorry, this was my first one so I got really confused. Should I close this one and create a fresh?

**Pull request closed without merging**